doc/release/1.15.0-notes.rst

```diff
@@ -287,6 +287,34 @@ of the overlap between input an output, that is, the next element accumulated
 is added before the accumulated result is stored in its place, hence the
 overlap is safe. Avoiding the copy results in faster execution.
 
+Generalized ufuncs now accept ``axes``, ``axis`` and ``keepdims`` arguments
+---------------------------------------------------------------------------
+One can control over which axes a generalized ufunc operates by passing in an
+``axes`` argument, a list of tuples with indices of particular axes.  For
+instance, for a signature of ``(i,j),(j,k)->(i,k)`` appropriate for matrix
+multiplication, the base elements are two-dimensional matrices and these are
+taken to be stored in the two last axes of each argument.  The corresponding
+axes keyword would be ``[(-2, -1), (-2, -1), (-2, -1)]``. If one wanted to
+use leading dimensions instead, one would pass in ``[(0, 1), (0, 1), (0, 1)]``.
+
+For simplicity, for generalized ufuncs that operate on 1-dimensional arrays
+(vectors), a single integer is accepted instead of a single-element tuple, and
+for generalized ufuncs for which all outputs are scalars, the (empty) output
+tuples can be omitted.  Hence, for a signature of ``(i),(i)->()`` appropriate
+for an inner product, one could pass in ``axes=[0, 0]`` to indicate that the
+vectors are stored in the first dimensions of the two inputs arguments.
+
+As a short-cut for generalized ufuncs that are similar to reductions, i.e.,
+that act on a single, shared core dimension such as the inner product example
+above, one can pass an ``axis`` argument. This is equivalent to passing in
+``axes`` with identical entries for all arguments with that core dimension
+(e.g., for the example above, ``axes=[(axis,), (axis,)]``).
```

Comment:
It would be nice to clarify (and test!) it this also works if the output has the same dimension. For example, a signature of `(i)->(i)`.

Reply:
Yes, indeed. The problem is that the `_umath_test` module does not contain any example of a gufunc with such a signature, and I don't feel quite up to writing one... (sadly, `linalg` also does not have one). Note, though, that the code path does get tested - with `keepdims=True`, any no-core arguments do get upgraded to having a single core dimension.

Reply:
p.s. Of course this isn't a reason to clarify the documentation! Indeed, the `keepdims` case itself might be useful for that.

Reply:
Now added a bit more text to `keepdims` to clarify this.